File: csharp/state-of-tic-tac-toe/StateOfTicTacToe.cs
public enum State
{
    Win,
    Draw,
    Ongoing,
    Invalid
}

public class TicTacToe
{

    string[] board;

    public TicTacToe(string[] rows)
    {
        if (rows.Count() != 3)
            throw new InvalidDataException("Board should have three rows.");
        if (rows.Any<string>(s => s.Length != 3))
            throw new InvalidDataException("Each row on the board should have three places.");

        board = rows;
    }

    private bool isWin(string[] board, string players = "XO")
    {
        string[] winners = (from char player in players select new String(player, 3)).ToArray<string>();
        bool anyRow = board.Any<string>(s => winners.Contains(s));
        bool anyCol = false;
        for (int i = 0; i < board[0].Length; i++)
        {
            foreach (char player in players)
            {
                anyCol = anyCol || (from row in board where row[i] == player select player).Count() == board.Length;
            }
        }
        bool diagonal = false;
        foreach (char player in players)
        {
            diagonal = diagonal ||
                (board[0][0] == player && board[1][1] == player && board[2][2] == player) ||
                (board[2][0] == player && board[1][1] == player && board[0][2] == player);
        }
        return anyRow || anyCol || diagonal;
    }

    private bool isFull(string[] board)
    {
        return (from row in board from ch in row where ch != 'X' && ch != 'O' select ch).Count() == 0;
    }

    private bool isDraw(string[] board)
    {
        return isFull(board) && !isWin(board);
    }

    private bool isWrongOrder(string[] board)
    {
        int countX = (from row in board from ch in row where ch == 'X' select ch).Count();
        int countY = (from row in board from ch in row where ch == 'O' select ch).Count();

        return countX != countY && countX != (countY + 1);
    }

    private bool playedAfterWin(string[] board)
    {
        return isWin(board, "X") && isWin(board, "O");
    }
    
    public State State
    {
        get
        {
            if (playedAfterWin(board) || isWrongOrder(board))
                return State.Invalid;
            else if (isWin(board))
                return State.Win;
            else if (isDraw(board))
                return State.Draw;
            else
                return State.Ongoing;
        }
    }
}

File: go/state-of-tic-tac-toe/state_of_tic_tac_toe.go
package stateoftictactoe

import (
	"errors"
	"strings"
)

type State string

const (
	Win     State = "win"
	Ongoing State = "ongoing"
	Draw    State = "draw"
	Invalid State = ""
)

// Determine game state for tic tac toe
// @param board: slice of strings of board state with a row for each string.
// @returns: State as "", "win", "ongoing", or "draw". "" only when there is an error.
// @raises: Error when players played out of order, or continued playing after a win
func StateOfTicTacToe(board []string) (State, error) {
	countX, countO := CountChar(board)
	if countX+countO == 0 { // No moves yet
		return Ongoing, nil
	}

	if countO > countX || countX > countO+1 {
		return Invalid, errors.New("Invalid player order")
	}
	xWin := IsWinner(board, 'X')
	oWin := IsWinner(board, 'O')
	if xWin && oWin || (xWin && countO >= countX) || (oWin && countX > countO) {
		// Kept playing after win
		return Invalid, errors.New("Invalid board")
	}
	if xWin || oWin {
		return Win, nil
	}
	if countX+countO == 9 {
		return Draw, nil
	}
	return Ongoing, nil
}

// Count the X's and O's on a tic-tac-toe board
// @param board: slice of string where each string is a row containing 'X', 'O', or ' ' (3 rows 3 columns assumed)
// @returns: Number of X's found, and Number of O's found.
func CountChar(board []string) (int, int) {
	x := 0
	o := 0
	for _, row := range board {
		for _, char := range row {
			switch char {
			case 'O':
				o++
			case 'X':
				x++
			}
		}
	}
	return x, o
}

// Check if a given player has won a game of tic-tac-toe
// @param board: Game board
// @param player: The player to check
// @param returns: True/False True if they have won.
func IsWinner(board []string, player rune) bool {
	targetRow := strings.Repeat(string(player), 3)
	for _, row := range []string{
		board[0],
		board[1],
		board[2],
		string(board[0][0]) + string(board[1][0]) + string(board[2][0]),
		string(board[0][1]) + string(board[1][1]) + string(board[2][1]),
		string(board[0][2]) + string(board[1][2]) + string(board[2][2]),
		string(board[0][0]) + string(board[1][1]) + string(board[2][2]),
		string(board[2][0]) + string(board[1][1]) + string(board[0][2]),
	} {
		if row == targetRow {
			return true
		}
	}
	return false
}
